tools/src/main/java/com/nedap/archie/json/flat/FlatJsonGenerator.java

```diff
@@ -102,6 +137,12 @@ private void buildPathsAndValuesInner(Map<String, Object> result, RMTypeInfo rmA
         }
     }
 
+    private boolean shouldAddTypeName(RMTypeInfo rmAttributeTypeInfo, OpenEHRBase rmObject, CObject cObject, boolean typeAlternativesPresent) {
+        return typeHasDescendants(rmAttributeTypeInfo) &&
+                !sameType(rmAttributeTypeInfo, rmObject) &&
+                ( !filterTypes || !sameType(cObject, rmObject) || typeAlternativesPresent);
+    }
```

Comment:
```suggestion
    /**
     * Determine if the type name attribute should be added.
     *
     * The type name attribute should be added if:
     *  - The actual RM object is of a different type then the RM attribute type; AND
     *  - One of the following conditions applies:
     *     - {@link FlatJsonFormatConfiguration#getFilterTypes()}  is disabled
     *     - The actual RM object is of a different type than specified in the archetype constraint
     *       (or no archetype constraint is available)
     *     - There are multiple alternative archetype constrains present.
     */
    private boolean shouldAddTypeName(RMTypeInfo rmAttributeTypeInfo, OpenEHRBase rmObject, CObject cObject, boolean typeAlternativesPresent) {
        return !sameType(rmAttributeTypeInfo, rmObject) &&
                ( !filterTypes || !sameType(cObject, rmObject) || typeAlternativesPresent);
    }
```

I think the first constraint is redundant and can be removed as the actual RM object can only be of a different type if the RM attribute type has descendants.